### ISSUE TITLE
sqlite: Customize display of numeric columns

### ISSUE THREAD

**Issue by mholt** [ultorg/public_issues#15]
First, this is amazing software and I'm quite excited about it. Thanks for sharing the beta with us; I hope I can keep using this for my personal projects.

I'm actually using the SQLite driver (issue #1 -- so far working like a charm!) and was wondering what you think about customizing the display of numeric columns. I know sqlite is weakly typed which might make this tricky, even though the columns are declared as integer or real.

Some columns contain UUIDs (their integer representations). Others contain unix-epoch timestamps. Yet others contain lat/lon coordinates (where the current truncation of the numbers are unacceptable; greater precision is needed).

Oh, and some are used as booleans (0 = false, 1 = true). It'd be nice to map columns to their semantic types so the display is more convenient.

Here's a sample of what it shows as right now:

![image](https://user-images.githubusercontent.com/1128849/167225047-2cd149ac-84e1-4326-bcec-7a3cd283023b.png)

Interestingly, the "Account" field should actually be the number 4, according to DBeaver. (I don't have UUIDs in this screenshot.)  I'm not sure if that's a bug or some other representation of an integer.

Maybe a right-click -> Format option for common interpretations of the number would be good! (Or if there's already a way to do that, just let me know, I am still learning this software.) Thanks

**Comment by eirikbakke** [comment]
You can adjust the precision of numbers with these two icons in the toolbar:
<img width="66" alt="precision" src="https://user-images.githubusercontent.com/886243/167227731-06f52ee8-fe62-41a6-9c7c-7d7d57ef80cc.png">

You can also click "Format" to open a list of formatting options for each field. But at the moment, the only numeric formatting options are the number of decimal digits and whether to group thousands or not.

For the Account field in your screenshot above, I'm wondering if the [Collapsed Join](https://github.com/ultorg/public_issues/issues/4) feature may be activated, leading to the confusing two numbers with a space between them. Can you right-click "Account", click "Fields & Join", and see if a "(Show Explicit Join)" option is available?

The official way to show an epoch timestamp as a real timestamp would be to use the `fromEpoch` formula function, though I don't think I have implemented it on SQLite yet. I could consider adding it as a formatting function, but it might be confusing to allow numbers to look like dates when there is already explicit timestamp types in the formula language.

A UUID is 128 bits or 16 bytes and so would have to be stored as either a BLOB or two INTEGER values in SQLite. Do you know which format is used in this particular database?

**Comment by mholt** [comment]
> You can adjust the precision of numbers with these two icons in the toolbar:

Brilliant, thanks!

> For the Account field in your screenshot above, I'm wondering if the https://github.com/ultorg/public_issues/issues/4 feature may be activated, leading to the confusing two numbers with a space between them. Can you right-click "Account", click "Fields & Join", and see if a "(Show Explicit Join)" option is available?

Indeed the Show Explicit Join option is available, and clicking it seemed to make more sense (I had to open a perspective on the table). After reading #4, I think I agree that while the collapsed join is cool and useful sometimes, I think seeing the explicit join (with the nested headers) is more obvious/clear and more interesting too.

> The official way to show an epoch timestamp as a real timestamp would be to use the fromEpoch formula function, though I don't think I have implemented it on SQLite yet. I could consider adding it as a formatting function, but it might be confusing to allow numbers to look like dates when there is already explicit timestamp types in the formula language.

So like `=fromEpoch([Timestamp])`? Seems you're right, there's an error (not sure how to copy the error message):

<img width="618" alt="Screen Shot 2022-05-06 at 10 51 43 PM" src="https://user-images.githubusercontent.com/1128849/167238893-ca4f1090-d466-4fa0-bc77-37b6944efb63.png">

> A UUID is 128 bits or 16 bytes and so would have to be stored as either a BLOB or two INTEGER values in SQLite. Is it a two-integer format that is used in this database?

Ah, you know, you're right, the only UUIDs stored here are as TEXT fields. Mea culpa.

Anyway, I realize this issue is very low-priority; I'm just starting using it, so I look forward to diving in more. (Alas, my primary DB is Sqlite for now.)

Thanks for the great work. :100:

**Comment by eirikbakke** [comment]
> So like =fromEpoch([Timestamp])?

Exactly! Yes, that's the error message I expected. I'll have to teach Ultorg how to translate the various formula functions on SQLite, so that the error goes away.
